src/passes/DeadArgumentElimination.cpp

```diff
@@ -176,49 +189,83 @@ struct DAE : public Pass {
   bool optimize = false;
 
   void run(Module* module) override {
+    DAEFunctionInfoMap infoMap;
+    // Ensure all entries exist so the parallel threads don't modify the data
+    // structure.
+    for (auto& func : module->functions) {
+      infoMap[func->name];
+    }
+    // The null name represents module-level code (not in a function).
+    infoMap[Name()];
+
     // Iterate to convergence.
     while (1) {
-      if (!iteration(module)) {
+      if (!iteration(module, infoMap)) {
         break;
       }
     }
   }
 
-  bool iteration(Module* module) {
+  bool iteration(Module* module, DAEFunctionInfoMap& infoMap) {
     allDroppedCalls.clear();
 
-    DAEFunctionInfoMap infoMap;
-    // Ensure they all exist so the parallel threads don't modify the data
-    // structure.
-    for (auto& func : module->functions) {
-      infoMap[func->name];
+#if DAE_DEBUG
+    // Enable this path to mark all contents as stale at the start of each
+    // iteration, which can be used to check for staleness bugs. Note, though,
+    // that staleness bugs can easily cause serious issues with validation (e.g.
+    // if data is stale we may miss that there is an additional caller, that
+    // prevents refining argument types etc.), so this may not be terribly
+    // helpful.
+    if (getenv("ALWAYS_MARK_STALE")) {
+      for (auto& [_, info] : infoMap) {
+        info.markStale();
+      }
     }
+#endif
+
     DAEScanner scanner(&infoMap);
     scanner.walkModuleCode(module);
-    for (auto& curr : module->exports) {
-      if (curr->kind == ExternalKind::Function) {
-        infoMap[curr->value].hasUnseenCalls = true;
-      }
-    }
     // Scan all the functions.
     scanner.run(getPassRunner(), module);
     // Combine all the info.
+    struct CallContext {
+      Call* call;
+      Function* func;
+    };
     std::map<Name, std::vector<Call*>> allCalls;
     std::unordered_set<Name> tailCallees;
-    for (auto& [_, info] : infoMap) {
+    std::unordered_set<Name> hasUnseenCalls;
+    // Track the function in which relevant expressions exist. When we modify
+    // those expressions we will need to mark the function's info as stale.
+    std::unordered_map<Expression*, Name> expressionFuncs;
+    for (auto& [func, info] : infoMap) {
       for (auto& [name, calls] : info.calls) {
         auto& allCallsToName = allCalls[name];
         allCallsToName.insert(allCallsToName.end(), calls.begin(), calls.end());
+        for (auto* call : calls) {
+          expressionFuncs[call] = func;
+        }
       }
       for (auto& callee : info.tailCallees) {
         tailCallees.insert(callee);
       }
-      for (auto& [name, calls] : info.droppedCalls) {
-        allDroppedCalls[name] = calls;
+      for (auto& [call, dropp] : info.droppedCalls) {
```

Comment:
Why two `p`s? Can we use `std::unordered_map<...>::merge` here?

Reply:
`dropp` is a `**` to a drop (as opposed to a `*`), which is a convention we use elsewhere. Maybe there's a better idea here?

Reply:
Ohh, got it. I didn't realize it was a pointer, and I was confused enough that I missed the hint, too.

Reply:
Would still be interesting to know if we can use `std::unordered_map<..>::merge`. I've never used that method, but it looks like it might do what you're doing manually here.

Reply:
Hmm, reading that now, despite the name, `merge` is a destructive operation on the source? That's... pretty scary :ghost: 

I suppose it *could* work here, as we don't need the previous data structure. However, this is a pattern we have in a lot of other places and it feels odd to me to use it specifically here, especially given this weird API.